base/compiler/ssair/ir.jl

```diff
@@ -1469,62 +1466,104 @@ function maybe_erase_unused!(
     return false
 end
 
-function fixup_phinode_values!(compact::IncrementalCompact, old_values::Vector{Any})
+struct FixedNode
+    node::Any
+    needs_fixup::Bool
+    FixedNode(@nospecialize(node), needs_fixup::Bool) = new(node, needs_fixup)
+end
+
+function fixup_phinode_values!(compact::IncrementalCompact, old_values::Vector{Any}, reify_new_nodes::Bool)
     values = Vector{Any}(undef, length(old_values))
+    needs_fixup = false
     for i = 1:length(old_values)
         isassigned(old_values, i) || continue
         val = old_values[i]
-        if isa(val, Union{OldSSAValue, NewSSAValue})
-            val = fixup_node(compact, val)
+        if isa(val, OldSSAValue)
+            val = compact.ssa_rename[val.id]
+            if isa(val, SSAValue)
+                compact.used_ssas[val.id] += 1
+            end
+        elseif isa(val, NewSSAValue)
+            if reify_new_nodes
+                val = SSAValue(length(compact.result) + val.id)
+            else
+                needs_fixup = true
+            end
         end
         values[i] = val
     end
-    values
+    return FixedNode(values, needs_fixup)
 end
 
-function fixup_node(compact::IncrementalCompact, @nospecialize(stmt))
+function fixup_node(compact::IncrementalCompact, @nospecialize(stmt), reify_new_nodes::Bool)
     if isa(stmt, PhiNode)
-        return PhiNode(stmt.edges, fixup_phinode_values!(compact, stmt.values))
+        (;node, needs_fixup) = fixup_phinode_values!(compact, stmt.values, reify_new_nodes)
+        return FixedNode(PhiNode(stmt.edges, node), needs_fixup)
     elseif isa(stmt, PhiCNode)
-        return PhiCNode(fixup_phinode_values!(compact, stmt.values))
+        (;node, needs_fixup) = fixup_phinode_values!(compact, stmt.values, reify_new_nodes)
+        return FixedNode(PhiCNode(node), needs_fixup)
     elseif isa(stmt, NewSSAValue)
-        return SSAValue(length(compact.result) + stmt.id)
-    elseif isa(stmt, OldSSAValue)
-        val = compact.ssa_rename[stmt.id]
-        if isa(val, SSAValue)
-            # If `val.id` is greater than the length of `compact.result` or
-            # `compact.used_ssas`, this SSA value is in `new_new_nodes`, so
-            # don't count the use
-            compact.used_ssas[val.id] += 1
+        if reify_new_nodes
+            return FixedNode(SSAValue(length(compact.result) + stmt.id), false)
+        else
+            return FixedNode(stmt, true)
         end
-        return val
+    elseif isa(stmt, OldSSAValue)
+        return FixedNode(compact.ssa_rename[stmt.id], false)
```

Comment:
Don't we need to fix the usage count here?
```suggestion
        val = compact.ssa_rename[stmt.id]
        if isa(val, SSAValue)
            compact.used_ssas[val.id] += 1
        end
        return FixedNode(compact.ssa_rename[stmt.id], false)
```

And if so, I think this part of `fixup_node` and `fixup_phinode_values!` should share the same code.